### PR TITLE
Support llama3

### PR DESCRIPTION
**Pull request by lvhan028**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: src/turbomind/kernels/attention/attention_params.h [hunk 1]
```diff
@@ -60,6 +60,11 @@ struct AttentionParams {
     float rotary_embedding_base;
     int   max_position_embeddings;
     float rope_ti_scale;  // used for linear RoPE scaling
```
Comment (AllentDan) [inline src/turbomind/kernels/attention/attention_params.h:62]: rope_ti_scale 参数跟  rope_scaling_factor 重复了吧

File: src/turbomind/models/llama/unified_attention_layer.cc [hunk 2]
```diff
@@ -238,12 +238,18 @@ inline void UnifiedAttentionLayer<T>::forward(TensorMap* outputs, const TensorMa
         params.rotary_embedding_dim    = params_.rotary_embedding_dim;
         params.rotary_embedding_base   = params_.rotary_embedding_base;
         params.max_position_embeddings = params_.max_position_embeddings;
+        params.rope_scaling_factor     = params_.rope_scaling_factor;
         params.rope_ti_scale           = 1.f;
-        if (!params_.use_dynamic_ntk && params_.rope_scaling_factor) {
+        if (params_.rope_scaling_type == "linear") {
             params.rope_ti_scale /= params_.rope_scaling_factor;
         }
```
Comment (AllentDan) [inline src/turbomind/models/llama/unified_attention_layer.cc:245]: 这里是不是可以去掉 rope_scaling_type 和 rope_ti_scale。array_ops.h 里直接用 high_freq_factor 做了区分

**Review by AllentDan — APPROVED** [review]
Tested llama3.1-8b-instruct OK.

File: src/turbomind/kernels/attention/array_ops.h [hunk 3]
```diff
@@ -251,6 +261,25 @@ struct FastRoPE {
         for (int i = 0; i < N; i += 2) {
             inv_freq_[i / 2] = ti_scale * exp2f((idx + i) * scale_factor);
         }
+        if (original_max_position_embeddings != 0) {
+            int old_context_len     = original_max_position_embeddings;
+            float low_freq_wavelen  = old_context_len / low_freq_factor;
+            float high_freq_wavelen = old_context_len / high_freq_factor;
```
Comment (lzhangzz) [inline src/turbomind/kernels/attention/array_ops.h:268]: These are constant to the kernel and should be pre-computed outside the kernel.

File: src/turbomind/kernels/attention/array_ops.h [hunk 4]
```diff
@@ -251,6 +261,25 @@ struct FastRoPE {
...
+            PRAGMA_UNROLL
+            for (int i = 0; i < N; i += 2) {
+                auto freq     = inv_freq_[i / 2];
+                float wavelen = 2 * M_PI / freq;
+                if (wavelen < high_freq_wavelen) {
+                    freq = freq;
+                } else if (wavelen > low_freq_wavelen) {
+                    freq = freq / scaling_factor;
```
Comment (lzhangzz) [inline src/turbomind/kernels/attention/array_ops.h:277]: precompute the inversion of `scaling_factor`

**Comment by lvhan028** [comment]
@zhulinJulia24 may add llama3.1 into tc

File: src/turbomind/kernels/attention/array_ops.h [hunk 5]
```diff
@@ -251,6 +261,25 @@ struct FastRoPE {
...
+                } else {
+                    auto smooth = (old_context_len / wavelen - low_freq_factor) / (high_freq_factor - low_freq_factor);
```
Comment (lzhangzz) [inline src/turbomind/kernels/attention/array_ops.h:280]: precompute the inversion of `high_freq_factor - low_freq_factor`

Comment (lzhangzz) [inline src/turbomind/kernels/attention/array_ops.h:280]:
```
 old_context_len / wavelen
= old_context_len / (2 * pi / freq)
= old_context_len / (2 * pi) * freq

=> old_context_len / (2 * pi) is constant to the kernel
```

**Comment by lzhangzz** [comment]
Simply try something like
```
inv_scaling_factor = 1 / scaling_factor;
old_ctx_len_div_2pi = old_context_len / (2 * PI)
inv_diff_freq_factor = 1 / (high_freq_factor - low_freq_factor);

auto smooth = (old_ctx_len_div_2pi * freq - low_freq_factor) * inv_diff_freq_factor;
smooth = max(0, min(1, smooth))
freq = (1 - smooth) * freq * inv_scaling_factor + smooth * freq;
```

File: src/turbomind/kernels/attention/array_ops.h [hunk 6]
```diff
@@ -251,6 +258,25 @@ struct FastRoPE {
...
+         * used by llama3.1 equals to the following equation, given the precommuted parameters as:
+        ```C++
+        inv_scaling_factor = 1 / factor;
+        inv_diff_freq_factor = 1 / (high_freq_factor - low_freq_factor);
+        alpha = old_context_len / (2 * PI) * inv_diff_freq_factor;
+        beta = low_freq_factor * inv_diff_freq_factor
+        ```
+        */
+        // clang-format on
+        if (llama3_inv_scaling_factor > 0.) {
```
Comment (lzhangzz) [inline src/turbomind/kernels/attention/array_ops.h:272]: 0. is a fp64 value

**Review by lzhangzz — APPROVED** [review]

**Pull request merged**